cpp/src/arrow/compute/kernels/boolean.cc

```diff
@@ -43,34 +43,35 @@ namespace compute {
 class InvertKernel : public UnaryKernel {
   Status Call(FunctionContext* ctx, const Datum& input, Datum* out) override {
     DCHECK_EQ(Datum::ARRAY, input.kind());
+    constexpr int64_t kZeroDestOffset = 0;
 
     const ArrayData& in_data = *input.array();
-    ArrayData* result;
-
-    out->value = ArrayData::Make(boolean(), in_data.length);
-    result = out->array().get();
+    std::shared_ptr<ArrayData> result = out->array();
+    result->type = boolean();
 
-    // Allocate or copy bitmap
+    // Handle validity bitmap
     result->null_count = in_data.null_count;
-    std::shared_ptr<Buffer> validity_bitmap = in_data.buffers[0];
+    const std::shared_ptr<Buffer>& validity_bitmap = in_data.buffers[0];
     if (in_data.offset != 0) {
-      RETURN_NOT_OK(CopyBitmap(ctx->memory_pool(), validity_bitmap->data(),
-                               in_data.offset, in_data.length, &validity_bitmap));
+      DCHECK_LE(BitUtil::BytesForBits(in_data.length), validity_bitmap->size());
+      CopyBitmap(validity_bitmap->data(), in_data.offset, in_data.length,
```

Comment:
The validity bitmap may not be allocated at all if the null count is 0, so this will crash